File: go/simplex/logger.go
package simplex

import (
	"fmt"
	"io"
	"os"
)

// LogLevel controls how much the SimplexLogger outputs.
type LogLevel int

const (
	Silent LogLevel = iota
	Info
	Debug
)

// SimplexLogger writes solver progress to a stream.
type SimplexLogger struct {
	Level  LogLevel
	stream io.Writer
}

// NewSilentLogger returns a logger that produces no output.
func NewSilentLogger() SimplexLogger {
	return SimplexLogger{Level: Silent, stream: io.Discard}
}

// NewLogger returns a logger writing to stdout at the given level.
func NewLogger(level LogLevel) SimplexLogger {
	return SimplexLogger{Level: level, stream: os.Stdout}
}

// NewFileLogger returns a logger writing to the given file path at the given level.
func NewFileLogger(path string, level LogLevel) (SimplexLogger, error) {
	f, err := os.Create(path)
	if err != nil {
		return SimplexLogger{}, err
	}
	return SimplexLogger{Level: level, stream: f}, nil
}

func (l SimplexLogger) write(s string) {
	if l.stream != nil {
		fmt.Fprintln(l.stream, s)
	}
}

// LogProblem logs the LP formulation if level >= Info.
func (l SimplexLogger) LogProblem(lp LinearProgram) {
	if l.Level == Silent {
		return
	}
	l.write("=======================")
	l.write(" Revised Simplex Solver")
	l.write("=======================")
	l.write(fmt.Sprintf("A = %v", lp.A))
	l.write(fmt.Sprintf("b = %v", lp.B))
	l.write(fmt.Sprintf("c = %v", lp.C))
	l.write("")
}

// LogPhase logs a phase header if level >= Info.
func (l SimplexLogger) LogPhase(phase int) {
	if l.Level == Silent {
		return
	}
	var header string
	if phase == 1 {
		header = "Phase 1 — Finding initial BFS"
	} else {
		header = "Phase 2 — Optimizing"
	}
	l.write(header)
}

// LogIteration logs iteration details if level == Debug.
func (l SimplexLogger) LogIteration(it int, basis BasisState, xb []float64, z float64) {
	if l.Level != Debug {
		return
	}
	l.write(fmt.Sprintf("  iter %d:", it))
	l.write(fmt.Sprintf("    xb       = %v", xb))
	l.write(fmt.Sprintf("    basic    = %v", basis.Basic))
	l.write(fmt.Sprintf("    nonbasic = %v", basis.Nonbasic))
	l.write(fmt.Sprintf("    z        = %v", z))
	l.write("")
}

// LogResult logs the solve result if level >= Info.
func (l SimplexLogger) LogResult(result SimplexResult) {
	if l.Level == Silent {
		return
	}
	l.write("")
	l.write(fmt.Sprintf("Result: %s", result.Status))
	l.write(fmt.Sprintf("  x          = %v", result.X))
	l.write(fmt.Sprintf("  z          = %v", result.Z))
	l.write(fmt.Sprintf("  iterations = %d", result.Iterations))
}

// Close closes the underlying stream if it is a file.
func (l SimplexLogger) Close() {
	if f, ok := l.stream.(*os.File); ok && f != os.Stdout && f != os.Stderr {
		f.Close()
	}
}

File: python/src/simplex/logger.py
from __future__ import annotations

import sys
from enum import Enum
from typing import IO

from .types import BasisState, LinearProgram, SimplexResult


class LogLevel(Enum):
    SILENT = "SILENT"
    INFO   = "INFO"
    DEBUG  = "DEBUG"


SILENT = LogLevel.SILENT
INFO   = LogLevel.INFO
DEBUG  = LogLevel.DEBUG


class SimplexLogger:
    def __init__(self, level: LogLevel = LogLevel.SILENT, stream: IO | None = None) -> None:
        self.level  = level
        self._stream = stream
        self._owns_stream = False

    @classmethod
    def from_path(cls, path: str, level: LogLevel = DEBUG) -> "SimplexLogger":
        logger = cls(level, open(path, "w"))
        logger._owns_stream = True
        return logger

    def _write(self, s: str) -> None:
        stream = self._stream if self._stream is not None else sys.stdout
        print(s, file=stream)

    def log_problem(self, lp: LinearProgram) -> None:
        if self.level == LogLevel.SILENT:
            return
        self._write("=======================")
        self._write(" Revised Simplex Solver")
        self._write("=======================")
        self._write(f"A =\n{lp.A}")
        self._write(f"b = {lp.b}")
        self._write(f"c = {lp.c}")
        self._write("")

    def log_phase(self, phase: int) -> None:
        if self.level == LogLevel.SILENT:
            return
        header = ("Phase 1 — Finding initial BFS" if phase == 1
                  else "Phase 2 — Optimizing")
        self._write(header)
        self._write("─" * len(header))

    def log_iteration(self, it: int, basis: BasisState,
                      xb: "np.ndarray", z: float) -> None:
        if self.level != LogLevel.DEBUG:
            return
        self._write(f"  iter {it}:")
        self._write(f"    xb       = {xb}")
        self._write(f"    basic    = {basis.basic}")
        self._write(f"    nonbasic = {basis.nonbasic}")
        self._write(f"    z        = {z}")
        self._write("")

    def log_result(self, result: SimplexResult) -> None:
        if self.level == LogLevel.SILENT:
            return
        self._write("")
        self._write(f"Result: {result.status.name}")
        self._write(f"  x          = {result.x}")
        self._write(f"  z          = {result.z}")
        self._write(f"  iterations = {result.iterations}")

    def close(self) -> None:
        if self._owns_stream and self._stream is not None:
            self._stream.close()
